### DISCUSSION TITLE
Adding modbus support to SKR 1.4 turbo

### DISCUSSION THREAD

**Discussion by ed-apex** [grblHAL/LPC176x#51] [Q&A]
Hi,

I recently got an SKR 3 EZ up and running and it's wonderful.  I have an SKR 1.4 turbo lying around which ended up not going into a printer project I did awhile back.  Looking at the board, it would seem that the pins on the wifi header, which can be mapped to be used as UART comm if USB_SERIAL_CDC is set to 0, could be setup to be used as a tx and rx for serial output to an rs485 transceiver.  The serial.c is significantly different in this repo than what's in Dresco's STM32H7xx repo.  I'm wondering if there's a reason I can't pilfer from that to try to define SERIAL_PORT_1 and use that for my spindle?  This is very much over my head at this point, but it seems like it should be possible.

Also... There's a disclaimer about ganged axis with this board.  It has sufficient axes to gang / auto square, I guess I'm not sure what's needing to be changed in the board map to make that happened.  With I haven't tested the fw yet, but it compiled fine with auto squaring enabled on the Y.

Sorry for the dumb questions, I super appreciate the help!

**Comment by terjeio** [comment]
Do you need two serial ports? If not then just use the one on the TFT header. If two then code has to be added for the second, best would be to copy the code from the current one and just use the code from the STM32 repo as a reference for how to modify it.

> Also... There's a disclaimer about ganged axis with this board.

Where?

**Reply by ed-apex** [reply]
I think I’m missing something. I’m using the skr 1.4 turbo for this project in the LPC176x repo. The boardmap doesn’t reference the TFT header anywhere. Do I just define SERIAL_PORT 1 in the boardmap? The serial.c in this repo lacks any of the conditional logic for setting up a serial port beyond the SERIAL_CDC and SERIAL_MOD. My understanding was that those were there to manage whether the normal board communications happen via serial over usb or via uart via the wifi header. 

The reference to the ganged axis limitation is in the boardmap footnote here https://github.com/grblHAL/LPC176x?tab=readme-ov-file#board-maps

**Reply by terjeio** [reply]
When SERIAL_CDC is set to 1 for a serial over USB connection the UART port is free for other use such as ModBus comms.

The UART serial port is by default routed to UART0, pin P0.2 & P0.3, which is connected to the TFT header according to the schematic. If you want to route it to the WIFI header instead just change `#if USB_SERIAL_CDC == 0` to `#if USB_SERIAL_CDC == 1` in _serial.c._

![image](https://github.com/user-attachments/assets/860d6113-ecee-49c2-9600-969cbe08e19a)

> The reference to the ganged axis limitation is in the boardmap footnote

Seems I forgot to update that when I added ganging/auto-squaring support...

**Comment by ed-apex** [comment]
So I should just be able to setup my board file for my huangyang spindle
and enable modbus along with enabling y auto squaring and sdcard? That’s
great!

Just so I can learn to understand the source a little better, can you
point me in the right direction as to where those pins are setup? I’m
trying to get a better handle on all of grblhal because I’m thinking this
is what I’m migrating all of my machines and future conversions too.

Also, I was only successful compiling this in MCUxpresso, do you have any
tips for building the platformio.ini file for a project that someone hasn’t
done this for yet?

Lastly, why is spindle synch a non-starter on this board? I’d love to throw
a quadrature encoder on my little Cnc lathe to make my teeny threaded parts
:)

Sorry for all the beginner questions.

**Reply by terjeio** [reply]
> So I should just be able to setup my board file for my huangyang spindle
and enable modbus along with enabling y auto squaring and sdcard? That’s
great!

ModBus will be enabled automatically when you enable a VFD like the Huanyang.

> Just so I can learn to understand the source a little better, can you
point me in the right direction as to where those pins are setup? I’m
trying to get a better handle on all of grblhal because I’m thinking this
is what I’m migrating all of my machines and future conversions too.

The serial port pins are set up [here](https://github.com/grblHAL/LPC176x/blob/0c731b05c93cf023399896cf690f2108ed99064a/src/serial.c#L337) - based on the configuration [here](https://github.com/grblHAL/LPC176x/blob/0c731b05c93cf023399896cf690f2108ed99064a/src/serial.c#L55-L58).

> Also, I was only successful compiling this in MCUxpresso, do you have any
tips for building the platformio.ini file for a project that someone hasn’t
done this for yet?

Info from a [net search](https://duckduckgo.com/?t=ffab&q=platformio+LPC1768&ia=web) combined with how platformio.ini is set up for a driver that has one may help?

> Lastly, why is spindle synch a non-starter on this board?

It _can_ be implemented if anyone decides to do so. For me it is a non-starter for several reasons:

* There is no FPU available, fixed-point artithmetic has to be implemented if the code from the other drivers is to be reused.
* It is an aging MCU that not many use for grblHAL.
* It will add to my workload and benefit few.

**Comment by ed-apex** [comment]
Thanks so much for the support! I’ll dig into making the platform.io ini a
bit more. I compiled in mcuxpresso with an added compiler flag for
ngc_expressions which worked. Obviously, I’m very much a novice /
enthusiast (ie far from pro) on the comp sci side of things.

Do you feel the newer STM32 btt boards are a good overall fit for grblhal
on an on going basis?

Also, thank you for your efforts to put this all together. I’ve done a
number of Mach 3 retrofits and this provides a cleaner and more cost
effective path forward. The changes in the landscape since I did my first
retrofit in the early 2000s when I was just out of college to now is
amazing and it’s due to the open source and enthusiasts community. I love
Mach 3 and am grateful for its existence and still use it literally
everyday, but this, to me is the future.

**Reply by terjeio** [reply]
> Do you feel the newer STM32 btt boards are a good overall fit for grblhal on an on going basis?

Definitely, the STM32H7xx boards are likely on par with the iMXRT1062 (Teensy 4.x) in terms of speed and may even offer better connectivity for those who need that. The STM32F4xx will also be with us for quite a while and is well supported by grblHAL.
The new RP2350 from Raspberry Pi is a good match for grblHAL, even more so than the RP2040. I expect to add spindle sync support for the RP2350 soon.

A while back I ditched Mach3/Smoothstepper ESS for grblHAL on a iMXRT1062 (with a T41U5XBB BOB) - resulting in a faster and smoother running machine. I have to admit I was kind of surprised by that...

**Reply by ed-apex** [reply]
That's funny, I was one of the test cases for the ethernet smoothstepper many moons ago when I converted my aged Multicam SF.  The Multicam board died and they didn't have any more anywhere.  It was one of my first conversion projects.  The one I'd really like to do is a point to point machine.  I'd need to find a board with a stupid number of outputs for the drill band solenoids, but the flexibility of the macro programming would make the calls for side boring way easier on a converted machine.  A lot of them use ONLY conversational controllers and not G Code (like old Biesse NC400 / 410) .  Great hardware, but painful to use.  You can steal them on the used market because of this.  A colleague just picked up a rover 20 in good working condtion for 2K!  There's one on FB in NH for 3 right now that looks minty too.  The spindle alone on one of those is worth the price of admission ;)

**Reply by ed-apex** [reply]
Sorry to keep bothering you, but for the life of me I can't get the firmware to initialize.  It compiles fine.  I took everything off the board, stripped it down to a my_machine.h that only has the board defined and the default settings for serial_cdc and no matter what I tweak or change it flashes, but a serial port device doesn't show up.  I changed the mcu in mcuexpresso to reflect the lpc1769 vs 68, but since it's doing the "external" build, I think that makes no difference and the system_core_clock where it's set in the default src is 120mhz.  I've tried manually forcing a 16k boot loader offset, didn't help.  Just out of ideas.

**Reply by terjeio** [reply]
It is caused by not having a required function (for modbus) in serial.c, try with this:

[serial.zip](https://github.com/user-attachments/files/18258313/serial.zip)

Sorry for the inconvenience caused.

**Comment by ed-apex** [comment]
Unfortunately, same result.  Even with a stripped down my_machine.h with
only the board definition (skr 4 turbo) and USB_SERIAL_CDC 1, it still
doesn't show up as a serial device.  I don't have a st link or usb to ttl
adapter or anything (both are on the way from amazon) so it's hard to
figure out what's happening.  I'm not super familiar with MCUXpresso, but I
believe I have it setup right.  I imported the source to the project with
copying the source into the work space.  I uncommented the post build lines
to make a .bin and away I went.  It compiles without error unless my many
failed experiments break something.  I made a point of starting a fresh
project with a fresh clone of the repo to test the new serial.src you sent
which I then copied in.  Sadly, no dice.  I spent way too long trying to
get a platformio implementation setup last night, but couldn't get my
linker script to work.  With the PIO registry of LPC1769 non existent, and
the board .json hacked up from a marlin / arduino implementation, I doubt
it would have solved the problem and more likely introduced more.  I'm just
not confident that I haven't done something wrong in setting up the project
in mcuxpresso.

**Reply by terjeio** [reply]
> I compiled in mcuxpresso with an added compiler flag for ngc_expressions which worked.

This means it compiled or compiled and ran successfully?

**Comment by ed-apex** [comment]
It compiled, I've not get it to run at all yet, or at least show up as a
serial port, without being able to interface with it via uart until the
adapters come, I won't know what's really happening.  I tried it without
any compiler flags or options in my_machine.h.  Windows just doesn't see
it.  To confirm the board itself is okay, I flashed marlin on there from a
repo for the skr 1.4 turbo and sure enough, shows up as com 4 in windows.
I'm missing something with grblhal and this board.

**Reply by terjeio** [reply]
Are you building the bootloader version? This is needed when programming via SD card.

![image](https://github.com/user-attachments/assets/3059fe2d-6b2c-4194-a535-31a585da4c7f)

**Comment by ed-apex** [comment]
I tried manually setting a 16kb offset in the memory linker, but I wasn't
sure if I did it right.  Is there a way to specify the offset
somewhere/someway else?

Edward Forer

Manager

401 486 6959

**Comment by ed-apex** [comment]
I just reread user.ld.  I just need to make a build config where it starts
with BL_?

**Comment by ed-apex** [comment]
Omfg. I just built the smoothie board bl version and it worked.
